Forward cookies if jsonData.keepCookies is specified in datasource config

18PatZ opened the pull request:
Grafana supports forwarding cookies to backend datasources if `keepCookies` is enabled for the datasource: https://github.com/grafana/grafana/pull/58132

This PR includes the cookies received from Grafana in the actual request. As an extra precaution, the request is only modified if `keepCookies` is enabled in the settings - otherwise there is no difference.

There should be no security concerns as Grafana passes _only_ the allowed cookies specified in `keepCookies` to `QueryDataRequest.Headers` (which becomes `requestHeaders` in the relevant code): https://github.com/grafana/grafana/pull/49541

Cookie forwarding is a necessity when Grafana is protected behind 3rd party authentication.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana-infinity-datasource?pullRequest=1258) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana-infinity-datasource?pullRequest=1258) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana-infinity-datasource?pullRequest=1258) it.</sub>

18PatZ commented on pkg/infinity/headers.go:
```diff
@@ -175,3 +175,15 @@ func getQueryReqHeader(requestHeaders map[string]string, headerName string) stri
+// For extra precaution, only forwards cookies if jsonData.keepCookies is set for the datasource
+// Grafana appears to provide only the specified cookies in the requestHeaders
+func ApplyForwardedCookies(_ context.Context, settings models.InfinitySettings, req *http.Request, requestHeaders map[string]string) *http.Request {
+	if len(settings.KeepCookies) >= 1 {
```
@yesoreyeram the setttings.KeepCookies is an array of strings pulled directly from `jsonData.keepCookies` in the datasource.yml configuration. That option is used automatically by Grafana to determine which cookies are populated into `QueryDataRequest.Headers`. If not specified, or empty, there will be no cookie in `QueryDataRequest.Headers`, and cookies outside of the ones specified in the array will not be present. 

So Grafana is handling the security of which cookies should be sent, and all this PR does is enable forwarding it onto the actual datasource. As such this if statement is not actually necessary, we can just forward the cookie header if it's present. I added it just to make doubly sure there will be no impact on users who haven't set the `jsonData.keepCookies` option (which currently has no affect on the Infinity plugin, so I doubt there are many).

18PatZ commented:
> Hi @18PatZ - Can you add the frontend part to enable cookie forwarding?

@yesoreyeram I've just added it as a configuration option under `URL, Headers & Params` (wasn't sure where exactly it should go). I also added a test datasource called `Postman Echo - Forward Cookies` which will show the `grafana_session_expiry` cookie in the echo response, if you are signed in to Grafana. You can make a new datasource and add/remove `grafana_session_expiry` and see that the response will not have the cookie if it's not set as an included cookie.

18PatZ commented:
> Code LGTM. Can you update the settings validation to include allowed hosts if there are more than 0 forwarded cookies found.
> 
> https://github.com/18PatZ/grafana-infinity-datasource/blob/7489d18260e3f80735c08d495bc0eb9dc77c8018/pkg/models/settings.go#L160

Done, it fails the health check now if include cookies is set without allowed hosts.

18PatZ commented:
Should I have bumped the package.json version or is that something you will handle?

18PatZ commented:
Thanks. Any idea when that will be? I'd just like to use the new changes as soon as I can 🙂
